model.py

```diff
@@ -336,7 +336,7 @@ def sessionmaker(cls, url=None):
         return sessionmaker(bind=engine)
 
     @classmethod
-    def initialize(cls, url):
+    def initialize(cls, url, create_materialized_work_class=True):
```

Comment:
What goes wrong if you don't have this option? I wrote nearly identical code recently but was able to remove it by changing the way materialized views are populated.

Reply:
I thought that indent looked familiar! In trying to create the class upon migration, an error was raised that the table didn't exist.

Reply:
Yeah, this should have been fixed. I encountered the same problem and the REFRESH MATERIALIZED VIEW in `initialize` forced the database session to recognize that the table existed. However, I wasn't able to figure out why the problem was happening in the first place, so you may have encountered a situation where the REFRESH doesn't fix the database session.

Reply:
Okay, let me finagle with it a bit more to work out what the exact error / problem is.

Reply:
I'm at a loss. I created a dev database with 100 works and removed all the materialized views. Then I ran the migration script and it created the materialized views and the data model object (which is where my error happened and I assume where yours happens), before complaining that there was no timestamp in the database.

Reply:
I got my database back to what I think was its previous state (??) without `materialized_work_for_lanes`, removed the `create_materialized_work_class` parameter from everywhere, and ran the migrations. I'm getting the opposite error from what I remember before--the materialized view is being created before the migration runs, so the migration generates an error. Here's the traceback:
```
2 new migrations found.
  - 20180105-2-materialized-work-for-lanes.sql
  - 20180109-add-workgenres-index.sql

ERROR: Migration has been halted.
/Users/courteneyervin/simplified/metadata/core/migration/20180105-2-materialized-work-for-lanes.sql must be migrated manually.
==================================================
Traceback (most recent call last):
  File "/Users/courteneyervin/simplified/metadata/core/scripts.py", line 2047, in run_migrations
    self._run_migration(full_migration_path, timestamp)
  File "/Users/courteneyervin/simplified/metadata/core/scripts.py", line 2087, in _run_migration
    self._db.execute(sql)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1154, in execute
    bind, close_with_result=True).execute(clause, params or {})
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 945, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
    context)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1402, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
    cursor.execute(statement, parameters)
ProgrammingError: (psycopg2.ProgrammingError) relation "mv_works_for_lanes" already exists
```

Reply:
I really could have sworn the error before was that the class couldn't be created without the materialized view.

Reply:
Okay, so that was the metadata database. I deleted `mv_works_for_lanes` in my circulation database, rolled back the `'Database Migration'` timestamp record to 2018-01-04, and tried to run migrations. This resulted in the error I expected:

```
Traceback (most recent call last):
  File "core/bin/migrate_database", line 11, in <module>
    DatabaseMigrationScript().run()
  File "/Users/courteneyervin/simplified/circulation/core/scripts.py", line 1895, in run
    self._session = SessionManager.session(url, initialize_data=False)
  File "/Users/courteneyervin/simplified/circulation/core/model.py", line 435, in session
    url
  File "/Users/courteneyervin/simplified/circulation/core/model.py", line 398, in initialize
    class MaterializedWorkWithGenre(Base, BaseMaterializedWork):
  File "/Users/courteneyervin/simplified/circulation/core/model.py", line 406, in MaterializedWorkWithGenre
    autoload_with=engine
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/sql/schema.py", line 440, in __new__
    metadata._remove_table(name, schema)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/sql/schema.py", line 435, in __new__
    table._init(name, metadata, *args, **kw)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/sql/schema.py", line 510, in _init
    self._autoload(metadata, autoload_with, include_columns)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/sql/schema.py", line 522, in _autoload
    self, include_columns, exclude_columns
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1972, in run_callable
    return conn.run_callable(callable_, *args, **kwargs)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1477, in run_callable
    return callable_(self, *args, **kwargs)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 364, in reflecttable
    return insp.reflecttable(table, include_columns, exclude_columns)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 563, in reflecttable
    table_name, schema, **table.dialect_kwargs):
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 369, in get_columns
    **kw)
  File "<string>", line 2, in get_columns
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 54, in cache
    ret = fn(self, con, *args, **kw)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2432, in get_columns
    info_cache=kw.get('info_cache'))
  File "<string>", line 2, in get_table_oid
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/engine/reflection.py", line 54, in cache
    ret = fn(self, con, *args, **kw)
  File "/Users/courteneyervin/simplified/circulation/env/lib/python2.7/site-packages/sqlalchemy/dialects/postgresql/base.py", line 2330, in get_table_oid
    raise exc.NoSuchTableError(table_name)
sqlalchemy.exc.NoSuchTableError: mv_works_for_lanes
```

Reply:
My follow-up question is, does this code from earlier in the method run, before the error happens?

```
            result = connection.execute(
                "REFRESH MATERIALIZED VIEW %s;" % view_name
            )
```
Judging from your line numbers, the code is there, and I don't see why it wouldn't run, so I think the issue might be that the REFRESH trick works on some versions of Postgres but not others, and your code should go in the branch.